jina/peapods/pods/compoundpod.py

```diff
@@ -0,0 +1,271 @@
+__copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+import argparse
+from argparse import Namespace
+from typing import Optional, Dict, List, Union, Set
+
+from .. import Pea
+from .. import Pod
+from .. import BasePod
+from ...enums import PollingType, PeaRoleType
+
+
+class CompoundPod(BasePod):
+    """A CompoundPod is a immutable set of pods, which run in parallel. They share the same input and output socket.
+    Internally, the peas of the pods can run with the process/thread backend. They can be also run in their own containers.
+    :param args: pod arguments parsed from the CLI
+    :param needs: pod names of preceding pods, the output of these pods are going into the input of this pod
+    """
+
+    def __init__(self, args: Union['argparse.Namespace', Dict], needs: Set[str] = None):
+        super().__init__(args, needs)
+        self.replica_list = []  # type: List['Pod']
+        if isinstance(args, Dict):
+            # This is used when a Pod is created in a remote context, where peas & their connections are already given.
+            self.replicas_args = args
+        else:
+            self.replicas_args = self._parse_args(args)
+
+    @property
+    def first_pod_args(self) -> Namespace:
+        """
+        Return the first non-head/tail pod args
+
+        :return: arguments of the first pod
+        """
+        # note this will be never out of boundary
+        return self.replicas_args['replicas'][0]
+
+    @property
+    def port_expose(self) -> int:
```

Comment:
This syntax is very confusing. Why is this needed?

It feels weird to ask a CompoundPod what is its port exposed

Reply:
I suggest thinking what properties need to be common and what not.

Needing this is a clear code smell.

Reply:
Well, I see now. It needs the `port_expose` of the `gateway` to know where to send `grpc requests`.

Reply:
But in this case, it is arbirary to take the `first replica`.

Reply:
I see your point. There will be a better solution for sure. 
Currently, the pod is using the same pattern. I fear that it is beyond the scope of this pr.
If you have an idea how to quickly change this I would go for this. If it requires too much time to refactor, let's do a followup pr.

Reply:
okey, but do the refactor PR soon after